modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/MultiEncodeWorkflowOperationHandler.java

```diff
@@ -0,0 +1,545 @@
+                + " does not match number of source Flavor segments " + sourceFlavors.length
+                + " (must be the same or one)");
+      }
+      n = sourceTags.length; // at least this many tracks
+    }
+    if (sourceFlavors != null) { // If flavors are used to select tracks
+      // If use source and target flavors, there should be the same number of them or all map into one target
+      if (targetFlavors != null && (targetFlavors.length != 1 && sourceFlavors.length != targetFlavors.length)) {
+        throw new WorkflowOperationException(
+                "number of source flavors " + sourceFlavors.length + " segment does not match number of target flavors"
+                        + targetFlavors.length + " (must be the same or one target flavor)");
+      }
+      // If use target tags, there should be the same number of source flavors and target tags or all map into one
+      // target tag
+      if (targetTags != null && (targetTags.length != 1 && sourceFlavors.length != targetTags.length)) {
+        throw new WorkflowOperationException(
+                "number of source flavors " + sourceFlavors.length + " segment does not match number of target Tags"
+                        + targetTags.length + " (must be the same or one target)");
+      }
+      // Number of profile groups should match number of source flavors
+      if ((profiles.length != 1 && sourceFlavors.length != profiles.length)) {
+        throw new WorkflowOperationException("number of source flavors segments " + sourceFlavors.length
+                + " does not match number of profiles segments " + profiles.length
+                + " (must be the same or one profile)");
+      }
+      if (sourceFlavors.length > n)
+        n = sourceFlavors.length; // at least this many tracks
+    }
+    int j = 0;
+    // One for each source flavor
+    for (int i = 0; i < n; i++) {
+      elementSelectors.add(new ElementProfileTagFlavor(profiles[j]));
+      if (profiles.length > 1)
+        j++; // All source use the same set of profiles or its own
+    }
+    // If uses tags to select, but sets target flavor, they must match
+    if (sourceTags != null && sourceFlavors != null) {
+      if (sourceTags.length != sourceFlavors.length && sourceFlavors.length != 1 && sourceTags.length != 1) {
+        throw new WorkflowOperationException(
+                "number of source flavors " + sourceTags.length + " does not match number of source tags "
+                        + sourceFlavors.length + " (must be the same or one set of tags or flavors)");
+      }
+    }
+    populateTag(elementSelectors, sourceTags);
+    populateFlavor(elementSelectors, sourceFlavors);
+    populateTargetTag(elementSelectors, targetTags);
+    populateTargetFlavor(elementSelectors, targetFlavors);
+    return elementSelectors;
+  }
+
+  private List<ElementProfileTagFlavor> populateTag(List<ElementProfileTagFlavor> elementSelectors, String[] sourceTags)
```

Comment:
This and the following three methods are very, very similar. Maybe you find a way to simplify this code? Seems quite redundant...